File: backend/app/models/monitoring.py
"""Pydantic models for rate limit monitoring config and status responses."""

from typing import Optional

from pydantic import BaseModel, field_validator


class MonitoringAccountConfig(BaseModel):
    """Per-account monitoring toggle."""

    enabled: bool = False


class MonitoringConfig(BaseModel):
    """Monitoring configuration stored in the settings table."""

    enabled: bool = False
    polling_minutes: int = 5
    accounts: dict[str, MonitoringAccountConfig] = {}

    @field_validator("polling_minutes")
    @classmethod
    def validate_polling_minutes(cls, v: int) -> int:
        allowed = {1, 5, 15, 30, 60}
        if v not in allowed:
            raise ValueError(f"polling_minutes must be one of {sorted(allowed)}")
        return v


class MonitoringConfigRequest(BaseModel):
    """Request body for POST /admin/monitoring/config."""

    enabled: bool = False
    polling_minutes: int = 5
    accounts: dict[str, MonitoringAccountConfig] = {}

    @field_validator("polling_minutes")
    @classmethod
    def validate_polling_minutes(cls, v: int) -> int:
        allowed = {1, 5, 15, 30, 60}
        if v not in allowed:
            raise ValueError(f"polling_minutes must be one of {sorted(allowed)}")
        return v


class ConsumptionRates(BaseModel):
    """Moving average consumption rates (tokens/hour) over different windows."""

    ten_min: Optional[float] = None
    thirty_min: Optional[float] = None
    sixty_min: Optional[float] = None
    one_eighty_min: Optional[float] = None


class EtaProjection(BaseModel):
    """Rate limit ETA projection result."""

    status: str  # "safe", "projected", "at_limit", "no_data"
    message: str
    eta: Optional[str] = None  # ISO datetime string
    minutes_remaining: Optional[float] = None
    resets_at: Optional[str] = None


class WindowSnapshot(BaseModel):
    """A single rate limit window snapshot with consumption rates and ETA."""

    account_id: int
    account_name: str = ""
    backend_type: str
    window_type: str
    tokens_used: int
    tokens_limit: int
    percentage: float
    threshold_level: str
    resets_at: Optional[str] = None
    recorded_at: str
    consumption_rates: ConsumptionRates
    eta: EtaProjection


class MonitoringStatusResponse(BaseModel):
    """Response for GET /admin/monitoring/status."""

    enabled: bool
    polling_minutes: int
    windows: list[WindowSnapshot]
    threshold_alerts: list[dict] = []


class SnapshotHistoryEntry(BaseModel):
    """A single entry in a snapshot history response."""

    tokens_used: int
    percentage: float
    recorded_at: str


class SnapshotHistoryResponse(BaseModel):
    """Response for GET /admin/monitoring/history."""

    account_id: int
    window_type: str
    history: list[SnapshotHistoryEntry]

File: frontend/src/services/api/types/monitoring.ts
/**
 * Monitoring, Backend, Account, and rate-limit types.
 */

export interface MonitoringAccountConfig {
  enabled: boolean;
}

export interface MonitoringConfig {
  enabled: boolean;
  polling_minutes: number;
  accounts: Record<string, MonitoringAccountConfig>;
}

export interface ConsumptionRates {
  '24h': number | null;
  '48h': number | null;
  '72h': number | null;
  '96h': number | null;
  '120h': number | null;
  unit?: string;  // '%/hr' for percentage-only mode, 'tok/hr' otherwise
}

export interface EtaProjection {
  status: 'safe' | 'projected' | 'at_limit' | 'no_data';
  message: string;
  eta: string | null;
  minutes_remaining: number | null;
  resets_at: string | null;
}

export interface WindowSnapshot {
  account_id: number;
  account_name: string;
  plan?: string;
  backend_type: string;
  window_type: string;
  tokens_used: number;
  tokens_limit: number;
  percentage: number;
  threshold_level: string;
  resets_at: string | null;
  recorded_at: string | null;
  consumption_rates: ConsumptionRates;
  eta: EtaProjection;
  shared_with?: string[];
  no_data?: boolean;
}

export interface MonitoringStatus {
  enabled: boolean;
  polling_minutes: number;
  windows: WindowSnapshot[];
  threshold_alerts: Array<{
    account_id: number;
    window_type: string;
    from_level: string;
    to_level: string;
  }>;
}

export interface SnapshotHistoryEntry {
  tokens_used: number;
  percentage: number;
  recorded_at: string;
}

export interface SnapshotHistory {
  account_id: number;
  window_type: string;
  history: SnapshotHistoryEntry[];
}

// AI Backends types
export interface BackendCapabilities {
  supports_json_output: boolean;
  supports_token_usage: boolean;
  supports_streaming: boolean;
  supports_non_interactive: boolean;
  json_output_flag?: string;
  non_interactive_flag?: string;
}

export interface AIBackend {
  id: string;
  name: string;
  type: string;  // 'claude', 'opencode', 'gemini', 'codex'
  description?: string;
  icon?: string;
  documentation_url?: string;
  is_installed: number;
  version?: string;
  models?: string[];  // Parsed from JSON
  capabilities?: BackendCapabilities;
  cli_path?: string;
  account_count?: number;
  account_emails?: string;  // Comma-separated emails from all accounts
  created_at?: string;
  last_used_at?: string;  // MAX(last_used_at) across all accounts
}

export interface RateLimitWindow {
  window_type: string;
  percentage: number;
  resets_at: string | null;
  tokens_used: number;
  tokens_limit: number;
}

export interface BackendAccount {
  id: number;
  backend_id: string;
  account_name: string;
  email?: string;
  config_path?: string;
  api_key_env?: string;
  is_default: number;
  plan?: string;
  usage_data?: Record<string, unknown>;
  created_at?: string;
}

export interface AIBackendWithAccounts extends AIBackend {
  accounts: BackendAccount[];
}
